// youtube.ts/dist/entities/Playlists.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
exports.__esModule = true;
var index_1 = require("./index");
var Playlists = /** @class */ (function () {
    function Playlists(api) {
        var _this = this;
        this.api = api;
        this.util = new index_1.Util(this.api);
        this.get = function (playlistResolvable, params) { return __awaiter(_this, void 0, void 0, function () {
            var id, response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!params)
                            params = {};
                        return [4 /*yield*/, this.util.resolveID(playlistResolvable, "playlist")];
                    case 1:
                        id = _a.sent();
                        params.id = id;
                        return [4 /*yield*/, this.api.get("playlist", params)];
                    case 2:
                        response = _a.sent();
                        return [2 /*return*/, response.items[0]];
                }
            });
        }); };
        this.items = function (playlistResolvable, params) { return __awaiter(_this, void 0, void 0, function () {
            var id, response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!params)
                            params = {};
                        return [4 /*yield*/, this.util.resolveID(playlistResolvable, "playlist")];
                    case 1:
                        id = _a.sent();
                        params.playlistId = id;
                        return [4 /*yield*/, this.api.get("playlistItems", params)];
                    case 2:
                        response = _a.sent();
                        return [2 /*return*/, response];
                }
            });
        }); };
        this.item = function (playlistResolvable, videoResolvable, params) { return __awaiter(_this, void 0, void 0, function () {
            var playlistId, videoId, response, i;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!params)
                            params = {};
                        return [4 /*yield*/, this.util.resolveID(playlistResolvable, "playlist")];
                    case 1:
                        playlistId = _a.sent();
                        return [4 /*yield*/, this.util.resolveID(videoResolvable, "video")];
                    case 2:
                        videoId = _a.sent();
                        params.playlistId = playlistId;
                        return [4 /*yield*/, this.api.get("playlistItems", params)];
                    case 3:
                        response = _a.sent();
                        for (i = 0; i < response.items.length; i++) {
                            if (response.items[i].contentDetails.videoId === videoId) {
                                return [2 /*return*/, response.items[i]];
                            }
                        }
                        return [2 /*return*/];
                }
            });
        }); };
        this.itemByID = function (playlistItemID, params) { return __awaiter(_this, void 0, void 0, function () {
            var response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!params)
                            params = {};
                        params.id = playlistItemID;
                        return [4 /*yield*/, this.api.get("playlistItems", params)];
                    case 1:
                        response = _a.sent();
                        if (!response.items[0])
                            return [2 /*return*/, Promise.reject("Invalid playlist item id.")];
                        return [2 /*return*/, response.items[0]];
                }
            });
        }); };
        this.search = function (params) { return __awaiter(_this, void 0, void 0, function () {
            var response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!params)
                            params = {};
                        params.type = "playlist";
                        return [4 /*yield*/, this.api.part("search", "snippet", params)];
                    case 1:
                        response = _a.sent();
                        return [2 /*return*/, response];
                }
            });
        }); };
    }
    return Playlists;
}());
exports.Playlists = Playlists;
