// youtube.ts/dist/youtube.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
var API_1 = __importDefault(require("./API"));
var entities_1 = require("./entities");
var Youtube = /** @class */ (function () {
    function Youtube(apiKey) {
        this.api = new API_1["default"](Youtube.apiKey);
        this.channels = new entities_1.Channels(this.api);
        this.videos = new entities_1.Videos(this.api);
        this.playlists = new entities_1.Playlists(this.api);
        this.comments = new entities_1.Comments(this.api);
        this.util = new entities_1.Util(this.api);
        this.info = new entities_1.Info(this.api);
        Youtube.apiKey = apiKey;
        this.api = new API_1["default"](Youtube.apiKey);
        this.channels = new entities_1.Channels(this.api);
        this.videos = new entities_1.Videos(this.api);
        this.util = new entities_1.Util(this.api);
        this.playlists = new entities_1.Playlists(this.api);
        this.comments = new entities_1.Comments(this.api);
        this.info = new entities_1.Info(this.api);
    }
    return Youtube;
}());
exports["default"] = Youtube;
module.exports["default"] = Youtube;
__export(require("./entities/index"));
